File: packages/react-native/android/src/main/java/com/hotupdater/DownloadService.kt
package com.hotupdater

import java.io.File
import java.net.URL

/**
 * Interface for custom download service implementations.
 *
 * Implement this interface to provide a custom networking stack for OTA bundle downloads.
 * This is useful for enterprise environments that require TLS pinning, corporate proxies,
 * or custom network interceptors.
 *
 * Example:
 * ```kotlin
 * class PinnedDownloadService : DownloadService {
 *     override suspend fun downloadFile(
 *         fileUrl: URL, destination: File,
 *         fileSizeCallback: ((Long) -> Unit)?,
 *         progressCallback: (Double) -> Unit
 *     ): DownloadResult {
 *         // Use your custom OkHttpClient with certificate pinning
 *     }
 * }
 *
 * // Before HotUpdater initializes:
 * HotUpdaterImpl.downloadServiceFactory = { PinnedDownloadService() }
 * ```
 */
interface DownloadService {
    /**
     * Downloads a file from a URL
     * @param fileUrl The URL to download from
     * @param destination The local file to save to
     * @param fileSizeCallback Optional callback called when file size is known
     * @param progressCallback Callback for download progress updates (0.0 to 1.0)
     * @return Result indicating success or failure
     */
    suspend fun downloadFile(
        fileUrl: URL,
        destination: File,
        fileSizeCallback: ((Long) -> Unit)? = null,
        progressCallback: (Double) -> Unit,
    ): DownloadResult
}

/**
 * Result wrapper for download operations
 */
sealed class DownloadResult {
    data class Success(
        val file: File,
    ) : DownloadResult()

    data class Error(
        val exception: Exception,
    ) : DownloadResult()
}

File: packages/react-native/ios/HotUpdater/Internal/DownloadService.swift
import Foundation

/**
 * Protocol for custom download service implementations.
 *
 * Implement this protocol to provide a custom networking stack for OTA bundle downloads.
 * This is useful for enterprise environments that require TLS pinning, corporate proxies,
 * or custom network interceptors.
 *
 * Example:
 * ```swift
 * class PinnedDownloadService: NSObject, DownloadService {
 *     func downloadFile(from url: URL, to destination: String, ...) -> URLSessionDownloadTask? {
 *         // Use your custom URLSession with certificate pinning
 *     }
 * }
 *
 * // Before HotUpdater initializes:
 * HotUpdaterImpl.downloadServiceFactory = { PinnedDownloadService() }
 * ```
 */
public protocol DownloadService {
    /**
     * Downloads a file from a URL.
     * @param url The URL to download from
     * @param destination The local path to save to
     * @param fileSizeHandler Optional callback called when file size is known
     * @param progressHandler Callback for download progress updates (0.0 to 1.0)
     * @param completion Callback with downloaded file URL or error
     * @return The download task, if applicable (may return nil for non-URLSession implementations)
     */
    func downloadFile(from url: URL, to destination: String, fileSizeHandler: ((Int64) -> Void)?, progressHandler: @escaping (Double) -> Void, completion: @escaping (Result<URL, Error>) -> Void) -> URLSessionDownloadTask?
}
